src/cmake.ts

```diff
@@ -422,6 +421,453 @@ export class ConfigurationReader {
+        const file = res[1];
+        const location = res[2];
+        const severity = res[3];
+        const code = res[4];
+        const message = res[5];
+        const abspath = path.isAbsolute(file)
+            ? file
+            : path.normalize(path.join(this.binaryDir, file));
+        const loc = (() => {
+            const parts = location.split(',');
+            if (parts.length === 1)
+                this.getTrimmedLineRange(file, Number.parseInt(parts[0]) - 1);
+            if (parseFloat.length === 2)
+                return new vscode.Range(
+                    Number.parseInt(parts[0]) - 1,
+                    Number.parseInt(parts[1]) - 1,
+                    Number.parseInt(parts[0]) - 1,
+                    Number.parseInt(parts[1]) - 1
+                );
+            if (parseFloat.length === 4)
+                return new vscode.Range(
+                    Number.parseInt(parts[0]) - 1,
+                    Number.parseInt(parts[1]) - 1,
+                    Number.parseInt(parts[2]) - 1,
+                    Number.parseInt(parts[3]) - 1
+                );
+            throw new Error('Unable to determine location of MSVC error');
+        })();
+        const diag = new vscode.Diagnostic(
+            loc,
+            message,
+            {
+                error: vscode.DiagnosticSeverity.Error,
+                warning: vscode.DiagnosticSeverity.Warning,
+                info: vscode.DiagnosticSeverity.Information,
+            }[severity]
+        );
+        diag.code = code;
+        diag.source = 'MSVC';
+        return [true, {
+            filepath: abspath,
+            key: full,
+            diag: diag,
+        }];
+    }
+}
+
+/**
+ * Parses a diagnostic message from Green Hills Compiler.
+ * Use single line error reporting when invoking GHS compiler (--no_wrap_diagnostics --brief_diagnostics).
+ *
+ * @class GHSDiagnosticParser
+ * @extends {DiagnosticParser}
+ */
+class GHSDiagnosticParser extends DiagnosticParser {
+    public parse(line: string): [boolean, Maybe<FileDiagnostic>] {
+        const diag = diagnostics.parseGHSDiagnostic(line);
+        if (!diag) {
+            return [false, null];
+        }
+        const abspath = path.isAbsolute(diag.file)
+            ? diag.file
+            : path.normalize(path.join(this.binaryDir, diag.file));
+        const vsdiag = new vscode.Diagnostic(
+            new vscode.Range(diag.line, diag.column, diag.line, diag.column),
+            diag.message,
+            {
+                error: vscode.DiagnosticSeverity.Error,
+                warning: vscode.DiagnosticSeverity.Warning,
+                remark: vscode.DiagnosticSeverity.Information,
+            }[diag.severity]
+        );
+        vsdiag.source = 'GHS';
+        return [true, {
+            filepath: abspath,
+            key: diag.full,
+            diag: vsdiag,
+        }];
+    }
+}
+
+interface IParserCollection {
+    [parser: string]: DiagnosticParser;
+}

```

Comment:
Instead of using an object to store things by strings, could we go with a `Map<string, DiagnosticParser>`? It is better checked by the compiler and better expresses intent at usage sites.